### PR TITLE
Adding support for contexts.

### LINKED ISSUE #8: Support multiple Grafana
Hi @safaci2000 
Can we support multiple Grafana like wizzy?
E.g: switch to Grafana dev
```
./bin/grafana-dashboard-manager set context grafana dev
```
Thank you,

### PR DESCRIPTION
**Pull request by safaci2000**
Fixes #8

BREAKING CHANGE:
  - Layout of the configuraiton has been changed, so previous
    configuration won't be compatible.
  - Saved imports of dashboards and datasources should not be affected.

### CODE REVIEW

File: api/user.go [hunk 1]
```diff
@@ -4,24 +4,36 @@ import (
...
+	"os"
 
+	"github.com/netsage-project/grafana-dashboard-manager/config"
 	"github.com/netsage-project/sdk"
 	"github.com/sirupsen/logrus"
 )
 
+func validateUserAPI(client *sdk.Client) {
+	if client == nil || !config.GetDefaultGrafanaConfig().AdminEnabled {
+		logrus.Info("Missing Admin client, please check your config and ensure basic auth is configured")
```
Comment (randolph-esnet) [inline api/user.go:16]: logrus has a Fatal method: https://pkg.go.dev/github.com/sirupsen/logrus#Fatal

File: api/user.go [hunk 2]
```diff
@@ -4,24 +4,36 @@ import (
...
+}
+
 //ListUsers list all grafana users
 func ListUsers(client *sdk.Client) []sdk.User {
 	ctx := context.Background()
+	validateUserAPI(client)
 	users, err := client.GetAllUsers(ctx)
 	if err != nil {
-		logrus.Panic("Unable to get users")
+		logrus.Error(err)
```
Comment (randolph-esnet) [inline api/user.go:27]: Same as above: logrus.Fatal

File: cmd/context.go [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+import (
+	"github.com/spf13/cobra"
+)
+
+var context = &cobra.Command{
+	Use:     "contexts",
+	Aliases: []string{"ctx", "context"},
+	Short:   "Manage Context configuration",
+	Long:    `Manage Context configuration which allows multiple grafana configs to be used.`,
```
Comment (randolph-esnet) [inline cmd/context.go:11]: nit: Make the quotes on this line consistent with the lines above. I suggest using ".

File: cmd/contextList.go [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+import (
+	"github.com/jedib0t/go-pretty/table"
+	"github.com/netsage-project/grafana-dashboard-manager/config"
+	"github.com/spf13/cobra"
+)
+
+var contextList = &cobra.Command{
+	Use:   "list",
+	Short: "List context",
+	Long:  `List contexts.`,
```
Comment (randolph-esnet) [inline cmd/contextList.go:12]: Same as above: use consistent quotes.

File: cmd/contextSet.go [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+import (
+	"github.com/netsage-project/grafana-dashboard-manager/config"
+	"github.com/spf13/cobra"
+)
+
+var contextSet = &cobra.Command{
+	Use:   "set",
+	Short: "set context",
+	Long:  `set contexts.`,
```
Comment (randolph-esnet) [inline cmd/contextSet.go:11]: Consistent quote.

File: cmd/contextShow.go [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+	"github.com/netsage-project/grafana-dashboard-manager/config"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	"gopkg.in/yaml.v2"
+)
+
+var contextShow = &cobra.Command{
+	Use:   "show",
+	Short: "show context",
+	Long:  `show contexts.`,
```
Comment (randolph-esnet) [inline cmd/contextShow.go:16]: Consistent quote

File: cmd/contextShow.go [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+	Run: func(cmd *cobra.Command, args []string) {
+		context, _ := cmd.Flags().GetString("context")
+		grafana := config.GetGrafanaConfig(context)
+		d, err := yaml.Marshal(grafana)
+		if err != nil {
+			logrus.Info("Failed to serialize context")
```
Comment (randolph-esnet) [inline cmd/contextShow.go:22]: Use logrus.Fatal

File: cmd/userPromote.go [hunk 8]
```diff
@@ -2,7 +2,9 @@ package cmd
 
 import (
 	"github.com/netsage-project/grafana-dashboard-manager/api"
+	"github.com/netsage-project/grafana-dashboard-manager/config"
 	"github.com/sirupsen/logrus"
+	log "github.com/sirupsen/logrus"
```
Comment (randolph-esnet) [inline cmd/userPromote.go:7]: This looks like a duplicate import. `log "github.com/sirupsen/logrus"` seems to be the more common pattern

File: cmd/contextShow.go [hunk 9]
```diff
@@ -0,0 +1,34 @@
+package cmd
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/netsage-project/grafana-dashboard-manager/config"
+	"github.com/sirupsen/logrus"
```
Comment (randolph-esnet) [inline cmd/contextShow.go:8]: Other files in this PR rename logrus to log. I suggest adding that here for consistency.

File: cmd/contextList.go [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+	Run: func(cmd *cobra.Command, args []string) {
+		tableObj.AppendHeader(table.Row{"context", "active"})
+		contexts := config.GetContexts()
+		activeContext := config.GetContext()
+		for _, item := range contexts {
```
Comment (randolph-esnet) [inline cmd/contextList.go:17]: nit: for _, item := range config.GetContexts()

Another nit: What about having config.GetContextMap() that returns the map[string]Context. Then you can range directly over the keys and not have to use the funk helper.

File: config/config.go [hunk 11]
```diff
@@ -1,62 +1,72 @@
...
 }
 
 func init() {
 	configData = &ConfigStruct{}
 	configData.defaultConfig = readViperConfig("importer")
-	grafana_config := configData.defaultConfig.GetStringMap("grafana")
-	grafana_yaml, _ := yaml.Marshal(grafana_config)
-	err := yaml.Unmarshal([]byte(grafana_yaml), &configData.grafanaConfig)
+	contexts := configData.defaultConfig.GetStringMap("contexts")
+	contextMaps, _ := yaml.Marshal(contexts)
```
Comment (randolph-esnet) [inline config/config.go:65]: Check this error and call log.Fatal if err != nil.

File: config/config.go [hunk 12]
```diff
@@ -1,62 +1,72 @@
...
+	err := yaml.Unmarshal([]byte(contextMaps), &configData.contextMap)
 	if err != nil {
-		panic(err)
+		log.Error("No valid configuration file has been found")
```
Comment (randolph-esnet) [inline config/config.go:68]: use log.Fatal

**Review by randolph-esnet — COMMENTED** [review]
I can't speak much for the business logic, but I do have some nitpicks and general style suggestions.

File: api/user.go (hunk 1 above)
Comment (safaci2000) [reply api/user.go:16]: I was trying to avoid that monster stack trace.  Does fatal just terminate after the message?

File: api/user.go (hunk 2 above)
Comment (safaci2000) [reply api/user.go:27]: Done

File: cmd/context.go (hunk 3 above)
Comment (safaci2000) [reply cmd/context.go:11]: The example code for cobra uses ` ` for the long description.  It's intended to be multiline I imagine, but fixed.

Comment (randolph-esnet) [reply cmd/context.go:11]: Ah. that makes sense. I'm on board with back-ticks then.

File: cmd/userPromote.go (hunk 8 above)
Comment (safaci2000) [reply cmd/userPromote.go:7]: Fixed.

File: cmd/contextShow.go (hunk 9 above)
Comment (safaci2000) [reply cmd/contextShow.go:8]: Done

File: cmd/contextList.go (hunk 10 above)
Comment (safaci2000) [reply cmd/contextList.go:17]: that feels like it's exposing too much internal data for no reason.  Plus funk I think provides some basic functionality that Go should just have IMO.

File: config/config.go (hunk 11 above)
Comment (safaci2000) [reply config/config.go:65]: Done

File: config/config.go (hunk 12 above)
Comment (safaci2000) [reply config/config.go:68]: Done

File: cmd/contextList.go (hunk 10 above)
Comment (randolph-esnet) [reply cmd/contextList.go:17]: 👍

**Review by randolph-esnet — APPROVED** [review]

**Pull request merged**